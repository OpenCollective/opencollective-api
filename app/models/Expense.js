const Temporal = require('sequelize-temporal');

module.exports = function(Sequelize, DataTypes) {

  var Expense = Sequelize.define('Expense', {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true
    },

    UserId: {
      type: DataTypes.INTEGER,
      references: 'Users',
      referencesKey: 'id',
      onDelete: 'SET NULL',
      onUpdate: 'CASCADE'
    },

    CollectiveId: {
      type: DataTypes.INTEGER,
      references: 'Groups',
      referencesKey: 'id',
      onDelete: 'SET NULL',
      onUpdate: 'CASCADE'
    },

    currency: {
      type: DataTypes.STRING,
      defaultValue: 'USD',
      set: function(val) {
        if (val && val.toUpperCase) {
          this.setDataValue('currency', val.toUpperCase());
        }
      }
    },

    amount: DataTypes.INTEGER,
    title: DataTypes.STRING,
    description: DataTypes.TEXT('long'),
    attachment: DataTypes.STRING,
    category: DataTypes.STRING,
    vat: DataTypes.INTEGER,

    LastEditedById: {
      type: DataTypes.INTEGER,
      references: 'Users',
      referencesKey: 'id',
      onDelete: 'SET NULL',
      onUpdate: 'CASCADE'
    },

    status: DataTypes.STRING,

    incurredAt: DataTypes.Date, // date when the expense was incurred

    createdAt: {
      type: DataTypes.DATE,
      defaultValue: Sequelize.NOW
    },

    updatedAt: {
      type: DataTypes.DATE,
      defaultValue: Sequelize.NOW
    },

    deletedAt: {
      type: DataTypes.DATE
    }
  }, {
    paranoid: true
  });

  return Temporal(Expense, Sequelize);
}
