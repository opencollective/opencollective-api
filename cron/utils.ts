import config from 'config';

/**
 * Heroku scheduler only has daily or hourly cron jobs, we only want to run
 * this script once per week on Monday (1). If the day is not Monday on production
 * we won't execute the script
 */
export function onlyExecuteInProdOnMondays() {
  const today = new Date();
  if (config.env === 'production' && today.getDay() !== 1) {
    console.log('OC_ENV is production and day is not Monday, script aborted!');
    process.exit(0);
  }
}
