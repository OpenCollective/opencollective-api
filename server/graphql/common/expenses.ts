import { ExpenseAttachment } from '../../models/ExpenseAttachment';

type ViewExpenseDataPermission = {
  attachments: boolean;
  payoutMethod: boolean;
  userLocation: boolean;
  invoiceInfo: boolean;
};

/**
 * Returns true if user is allowed to see the private infos for an expense.
 * Based on loaders, so it's safe to use this function in lists.
 */
export const canViewExpensePrivateInfo = async (expense, req): Promise<ViewExpenseDataPermission> => {
  if (!req.remoteUser) {
    return {
      attachments: false,
      payoutMethod: false,
      userLocation: false,
      invoiceInfo: false,
    };
  } else if (req.remoteUser.isAdmin(expense.FromCollectiveId) || req.remoteUser.id === expense.UserId) {
    // Users can see the all the private info for the collective they're admin of and the expenses they submit
    return {
      attachments: true,
      payoutMethod: true,
      userLocation: true,
      invoiceInfo: true,
    };
  } else if (req.remoteUser.isAdmin(expense.CollectiveId)) {
    // Collective admins can see the attachments only
    return {
      attachments: true,
      payoutMethod: false,
      userLocation: false,
      invoiceInfo: false,
    };
  } else {
    // Users can see the private info for the expenses submitted to collectives they're hosting
    const collective = await req.loaders.Collective.byId.load(expense.CollectiveId);
    const isHostAdmin = req.remoteUser.isAdmin(collective.HostCollectiveId);
    const isParentCollectiveAdmin = req.remoteUser.isAdmin(collective.ParentCollectiveId);
    return {
      attachments: isHostAdmin || isParentCollectiveAdmin,
      payoutMethod: isHostAdmin,
      userLocation: isHostAdmin,
      invoiceInfo: isHostAdmin,
    };
  }
};

/**
 * Returns the list of attachments for this expense.
 */
export const getExpenseAttachments = async (expenseId, req): Promise<ExpenseAttachment[]> => {
  return req.loaders.ExpenseAttachment.byExpenseId.load(expenseId);
};
