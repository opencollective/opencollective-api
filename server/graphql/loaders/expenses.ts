import DataLoader from 'dataloader';
import { groupBy } from 'lodash';

import ACTIVITY from '../../constants/activities';
import { TransactionKind } from '../../constants/transaction-kind';
import queries from '../../lib/queries';
import models, { Op, sequelize } from '../../models';
import { Activity } from '../../models/Activity';
import { ExpenseAttachedFile } from '../../models/ExpenseAttachedFile';
import { ExpenseItem } from '../../models/ExpenseItem';
import { LEGAL_DOCUMENT_TYPE } from '../../models/LegalDocument';

import { sortResultsArray } from './helpers';

/**
 * Loader for expense's items.
 */
export const generateExpenseItemsLoader = (): DataLoader<number, ExpenseItem[]> => {
  return new DataLoader(async (expenseIds: number[]) => {
    const items = await models.ExpenseItem.findAll({
      where: { ExpenseId: { [Op.in]: expenseIds } },
      order: [['id', 'ASC']],
    });

    return sortResultsArray(expenseIds, items, item => item.ExpenseId);
  });
};

/**
 * Load all activities for an expense
 */
export const generateExpenseActivitiesLoader = (): DataLoader<number, Activity[]> => {
  return new DataLoader(async (expenseIDs: number[]) => {
    const activities = await models.Activity.findAll({
      order: [['createdAt', 'ASC']],
      where: {
        ExpenseId: {
          [Op.in]: expenseIDs,
        },
        type: {
          [Op.in]: [
            ACTIVITY.COLLECTIVE_EXPENSE_CREATED,
            ACTIVITY.COLLECTIVE_EXPENSE_DELETED,
            ACTIVITY.COLLECTIVE_EXPENSE_UPDATED,
            ACTIVITY.COLLECTIVE_EXPENSE_INVITE_DRAFTED,
            ACTIVITY.COLLECTIVE_EXPENSE_REJECTED,
            ACTIVITY.COLLECTIVE_EXPENSE_APPROVED,
            ACTIVITY.COLLECTIVE_EXPENSE_MOVED,
            ACTIVITY.COLLECTIVE_EXPENSE_UNAPPROVED,
            ACTIVITY.COLLECTIVE_EXPENSE_PAID,
            ACTIVITY.COLLECTIVE_EXPENSE_MARKED_AS_UNPAID,
            ACTIVITY.COLLECTIVE_EXPENSE_PROCESSING,
            ACTIVITY.COLLECTIVE_EXPENSE_ERROR,
            ACTIVITY.COLLECTIVE_EXPENSE_SCHEDULED_FOR_PAYMENT,
            ACTIVITY.COLLECTIVE_EXPENSE_MARKED_AS_SPAM,
            ACTIVITY.COLLECTIVE_EXPENSE_MARKED_AS_INCOMPLETE,
          ],
        },
      },
    });

    return sortResultsArray(expenseIDs, activities, activity => activity.ExpenseId);
  });
};

/**
 * Loader for expense's attachedFiles.
 */
export const attachedFiles = (): DataLoader<number, ExpenseAttachedFile[]> => {
  return new DataLoader(async (expenseIds: number[]) => {
    const attachedFiles = await models.ExpenseAttachedFile.findAll({
      where: { ExpenseId: { [Op.in]: expenseIds } },
    });

    return sortResultsArray(expenseIds, attachedFiles, file => file.ExpenseId);
  });
};

/**
 * Expense loader to check if userTaxForm is required before expense payment
 */
export const userTaxFormRequiredBeforePayment = (): DataLoader<number, boolean> => {
  return new DataLoader<number, boolean>(async (expenseIds: number[]): Promise<boolean[]> => {
    const expenseIdsPendingTaxForm = await queries.getTaxFormsRequiredForExpenses(expenseIds);
    return expenseIds.map(id => expenseIdsPendingTaxForm.has(id));
  });
};

/**
 * Loader for expense's requiredLegalDocuments.
 */
export const requiredLegalDocuments = (): DataLoader<number, string[]> => {
  return new DataLoader(async (expenseIds: number[]) => {
    const expenseIdsPendingTaxForm = await queries.getTaxFormsRequiredForExpenses(expenseIds);
    return expenseIds.map(id => (expenseIdsPendingTaxForm.has(id) ? [LEGAL_DOCUMENT_TYPE.US_TAX_FORM] : []));
  });
};

/**
 * Should only be used with paid expenses
 */
export const generateExpenseToHostTransactionFxRateLoader = (): DataLoader<
  number,
  { rate: number; currency: string }
> =>
  new DataLoader(async (expenseIds: number[]) => {
    const transactions = await models.Transaction.findAll({
      raw: true,
      attributes: ['ExpenseId', 'currency', [sequelize.json('data.expenseToHostFxRate'), 'expenseToHostFxRate']],
      where: {
        ExpenseId: expenseIds,
        kind: TransactionKind.EXPENSE,
        type: 'CREDIT',
        isRefund: false,
        RefundTransactionId: null,
        data: { expenseToHostFxRate: { [Op.ne]: null } },
      },
    });

    const groupedTransactions = groupBy(transactions, 'ExpenseId');
    return expenseIds.map(expenseId => {
      const transactionData = groupedTransactions[expenseId]?.[0];
      const rate = parseFloat(transactionData?.expenseToHostFxRate);
      return isNaN(rate) ? null : { rate, currency: transactionData?.currency };
    });
  });
