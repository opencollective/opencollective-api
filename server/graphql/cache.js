import { difference, isEqual } from 'lodash';

import { TransactionKind } from '../constants/transaction-kind';
import { md5 } from '../lib/utils';

export const purgeCacheForCollectiveOperationNames = [
  'CollectivePage',
  'ContributePage',
  'BudgetSection',
  'UpdatesSection',
  'TransactionsSection',
  'TransactionsPage',
  'ExpensesPage',
  'CollectiveBannerIframe',
  'CollectiveCover',
  'RecurringContributions',
  'ContributionsSection',
  'Members_Users',
  'Members_Organizations',
];

export function checkSupportedVariables(req, variableNames) {
  if (!req.body.variables) {
    return false;
  }
  // Check missing variables
  if (difference(variableNames, Object.keys(req.body.variables)).length > 0) {
    return false;
  }
  // Check extra variables
  if (difference(Object.keys(req.body.variables), variableNames).length > 0) {
    return false;
  }

  return true;
}

export function getGraphqlCacheKey(req) {
  if (req.remoteUser) {
    return;
  }
  if (!req.body || !req.body.operationName || !req.body.query) {
    return;
  }

  const queryHash = md5(req.body.query);

  switch (req.body.operationName) {
    case 'CollectivePage':
    case 'ContributePage':
      if (!checkSupportedVariables(req, ['slug', 'nbContributorsPerContributeCard'])) {
        return;
      }
      if (req.body.variables.nbContributorsPerContributeCard !== 4) {
        return;
      }
      return `${req.body.operationName}_${queryHash}_${req.body.variables.slug}`;
    case 'BudgetSection':
      if (!checkSupportedVariables(req, ['slug', 'limit', 'hostSlug', 'kind'])) {
        return;
      }
      if (req.body.variables.limit !== 3) {
        return;
      }
      if (
        req.body.variables.kinds &&
        !isEqual(req.body.variables.kinds.sort(), [
          TransactionKind.ADDED_FUNDS,
          TransactionKind.CONTRIBUTION,
          TransactionKind.EXPENSE,
          TransactionKind.PLATFORM_TIP,
        ])
      ) {
        return;
      }
      return `${req.body.operationName}_${queryHash}_${req.body.variables.slug}`;
    case 'UpdatesSection':
      if (!checkSupportedVariables(req, ['slug', 'onlyPublishedUpdates'])) {
        return;
      }
      if (req.body.variables.onlyPublishedUpdates !== true) {
        return;
      }
      return `${req.body.operationName}_${queryHash}_${req.body.variables.slug}`;
    case 'TransactionsSection':
      if (!checkSupportedVariables(req, ['slug', 'limit'])) {
        return;
      }
      if (req.body.variables.limit !== 10) {
        return;
      }
      return `${req.body.operationName}_${queryHash}_${req.body.variables.slug}`;
    case 'TransactionsPage':
      if (!checkSupportedVariables(req, ['slug', 'offset', 'limit'])) {
        return;
      }
      if (req.body.variables.offset !== 0 || req.body.variables.limit !== 15) {
        return;
      }
      return `${req.body.operationName}_${queryHash}_${req.body.variables.slug}`;
    case 'ExpensesPage':
      if (!checkSupportedVariables(req, ['collectiveSlug', 'offset', 'limit'])) {
        return;
      }
      if (req.body.variables.offset !== 0 || req.body.variables.limit !== 10) {
        return;
      }
      return `${req.body.operationName}_${queryHash}_${req.body.variables.collectiveSlug}`;
    case 'CollectiveBannerIframe':
      if (!checkSupportedVariables(req, ['collectiveSlug'])) {
        return;
      }
      return `${req.body.operationName}_${queryHash}_${req.body.variables.collectiveSlug}`;
    case 'CollectiveCover':
    case 'RecurringContributions':
    case 'ContributionsSection':
      if (!checkSupportedVariables(req, ['slug'])) {
        return;
      }
      return `${req.body.operationName}_${queryHash}_${req.body.variables.slug}`;
    case 'Members':
      if (!checkSupportedVariables(req, ['collectiveSlug', 'offset', 'limit', 'type', 'role', 'orderBy'])) {
        return;
      }
      if (req.body.variables.offset !== 0 || req.body.variables.limit !== 100) {
        return;
      }
      if (req.body.variables.role !== 'BACKER') {
        return;
      }
      if (req.body.variables.type === 'ORGANIZATION,COLLECTIVE') {
        return `${req.body.operationName}_${queryHash}_Organizations_${req.body.variables.collectiveSlug}`;
      }
      if (req.body.variables.type === 'USER') {
        return `${req.body.operationName}_${queryHash}_Users_${req.body.variables.collectiveSlug}`;
      }
  }
}
