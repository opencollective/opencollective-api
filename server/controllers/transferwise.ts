import config from 'config';
import { Request, Response } from 'express';

import expenseStatus from '../constants/expense-status';
import { getExpenseFees, setTransferWiseExpenseAsProcessing } from '../graphql/common/expenses';
import { idDecode, IDENTIFIER_TYPES } from '../graphql/v2/identifiers';
import errors from '../lib/errors';
import logger from '../lib/logger';
import { reportErrorToSentry, reportMessageToSentry } from '../lib/sentry';
import { simulateTransferSuccess } from '../lib/transferwise';
import models, { Op } from '../models';
import transferwise from '../paymentProviders/transferwise';
import { handleTransferStateChange } from '../paymentProviders/transferwise/webhook';
import { BatchGroup } from '../types/transferwise';

const processPaidExpense = (host, remoteUser, batchGroup: BatchGroup) => async expense => {
  try {
    await expense.reload();
    if (expense.data?.transfer) {
      const payoutMethod = await expense.getPayoutMethod();
      const { feesInHostCurrency } = await getExpenseFees(expense, host, { payoutMethod });
      return setTransferWiseExpenseAsProcessing({
        expense,
        host,
        data: { batchGroup },
        feesInHostCurrency,
        remoteUser,
      });
    }
  } catch (e) {
    logger.error(`Wise Batch Payment: Error processing paid expense ${expense.id}`, e);
  }
};

export async function payBatch(
  req: Request<any, any, { expenseIds: Array<string>; hostId: string }>,
  res: Response,
): Promise<void> {
  try {
    const { remoteUser, headers, body } = req;
    if (!remoteUser) {
      throw new errors.Unauthorized('User needs to be logged in');
    }

    const host = await models.Collective.findByPk(idDecode(body.hostId, IDENTIFIER_TYPES.ACCOUNT));
    if (!host) {
      throw new errors.NotFound('Could not find host collective');
    }
    if (!remoteUser.isAdmin(host.id)) {
      throw new errors.Unauthorized('User must be admin of host collective');
    }
    const expenseIds = body.expenseIds?.map(id => idDecode(id, IDENTIFIER_TYPES.EXPENSE));
    const expenses = await models.Expense.findAll({
      where: { id: { [Op.in]: expenseIds } },
      include: [
        { model: models.PayoutMethod, as: 'PayoutMethod', required: true },
        {
          model: models.Collective,
          as: 'collective',
          attributes: [],
          // TODO: We should ideally use the host attached to the expense. See https://github.com/opencollective/opencollective/issues/4271
          where: { HostCollectiveId: host.id },
          required: true,
        },
      ],
    });

    expenses.forEach(expense => {
      if (expense.status !== expenseStatus.SCHEDULED_FOR_PAYMENT) {
        throw new Error(`Expense ${expense.id} must be scheduled for payment`);
      }
    });

    if (expenseIds.length !== expenses.length) {
      const errorInfo = { requested: expenseIds, found: expenses.map(e => e.id) };
      logger.error(`Wise Batch Pay: Could not find all requested expenses. ${JSON.stringify(errorInfo)}`);
      reportMessageToSentry('Wise Batch Pay: Could not find all requested expenses', { extra: errorInfo });
      throw new errors.NotFound('Could not find requested expenses');
    }
    const ottHeader = headers['x-2fa-approval'] as string;

    const fundResponse = ottHeader
      ? // Forward OTT response if included
        await transferwise.payExpensesBatchGroup(host, undefined, ottHeader, remoteUser)
      : // Otherwise, send the list of Expenses to pay the batch
        await transferwise.payExpensesBatchGroup(host, expenses, undefined, remoteUser);

    // If OTT response, proxy it to the frontend and return early
    if ('status' in fundResponse && 'headers' in fundResponse) {
      res.setHeader('x-2fa-approval', fundResponse.headers['x-2fa-approval']);
      res.sendStatus(fundResponse.status);
    } else if (fundResponse.status === 'COMPLETED') {
      // Mark expenses as paid and create transactions
      await Promise.all(expenses.map(processPaidExpense(host, remoteUser, fundResponse)));
      // Send 200 to the frontend
      res.sendStatus(200);
      // Simulate transfer success in other environments so transactions don't get stuck.
      if (['development', 'staging'].includes(config.env)) {
        const connectedAccount = await host.getAccountForPaymentProvider('transferwise');
        logger.debug(`Wise: Simulating transfer success for batch group ${fundResponse.id}`);
        for (const transferId of fundResponse.transferIds) {
          const response = await simulateTransferSuccess(connectedAccount, transferId);
          logger.debug(`Wise: Simulated transfer success for transfer ${transferId}`);
          const expense = expenses.find(e => e.data.transfer.id === transferId);
          await expense.update({ data: { ...expense.data, transfer: response } });
          // In development mode we don't have webhooks set up, so we need to manually trigger the event handler.
          if (config.env === 'development') {
            await handleTransferStateChange({
              // eslint-disable-next-line camelcase
              data: { resource: response, current_state: 'outgoing_payment_sent' },
            } as any);
          }
        }
      }
    } else {
      throw new Error(`Could not pay for batch group #${fundResponse.id}, please contact support.`);
    }
  } catch (e) {
    logger.error('Error paying Wise batch group', e);
    reportErrorToSentry(e);
    res
      .status(e.code || 500)
      .send(e.toString())
      .end();
  }
}
