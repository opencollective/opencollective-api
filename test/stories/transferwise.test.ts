/* eslint-disable camelcase */

import { expect } from 'chai';
import config from 'config';
import { round } from 'lodash';

import ExpenseStatuses from '../../server/constants/expense_status';
import { payExpense } from '../../server/graphql/common/expenses';
import cache from '../../server/lib/cache';
import models from '../../server/models';
import Expense from '../../server/models/Expense';
import { PayoutMethodTypes } from '../../server/models/PayoutMethod';
import { handleTransferStateChange } from '../../server/paymentProviders/transferwise/webhook';
import {
  fakeCollective,
  fakeConnectedAccount,
  fakeExpense,
  fakePayoutMethod,
  fakeTransaction,
  fakeUser,
} from '../test-helpers/fake-data';
import { resetTestDB, snapshotLedger, useIntegrationTestRecorder } from '../utils';

describe('/test/stories/transferwise.test.ts', () => {
  useIntegrationTestRecorder(config.transferwise.apiUrl, __filename, nock => {
    // Ignore our randomly generated customerTransactionId
    if (nock.body?.customerTransactionId) {
      nock.body.customerTransactionId = /.+/i;
    }
    return nock;
  });

  before(async () => {
    await resetTestDB();
    await cache.clear();
    config.fixer = { accessKey: 'fake-token' };
  });

  const setupTest = async ({ expenseCurrency, collectiveCurrency, payoutData }) => {
    await models.Transaction.truncate();
    const hostCurrency = 'USD';
    const expenseAmount = 100e2;
    const hostAdmin = await fakeUser();
    const host = await fakeCollective({
      admin: hostAdmin.collective,
      plan: 'network-host-plan',
      currency: hostCurrency,
      name: 'Open Source Collective',
    });
    const collective = await fakeCollective({
      HostCollectiveId: host.id,
      currency: collectiveCurrency,
      name: `Babel ${collectiveCurrency}`,
    });
    const user = await fakeUser(undefined, { name: 'Donnortello' });
    await fakeConnectedAccount({
      CollectiveId: host.id,
      service: 'transferwise',
      token: '3a0758c0-1df1-4995-91ee-21fb56a2a24b',
      data: {
        type: 'personal',
        id: 6220,
        userId: 5466594,
        address: {
          addressFirstLine: '56 Shoreditch High Street',
          city: 'London',
          countryIso2Code: 'GB',
          countryIso3Code: 'gbr',
          postCode: 'E16JJ',
          stateCode: null,
        },
        email: '',
        createdAt: '2020-01-21T18:13:48.000Z',
        updatedAt: '2020-07-28T11:55:53.000Z',
        obfuscated: false,
        currentState: 'VISIBLE',
        firstName: 'Leo',
        lastName: 'Kewitz',
        dateOfBirth: '1964-02-24',
        phoneNumber: '+442038087139',
        secondaryAddresses: [],
        fullName: 'Leo Kewitz',
      },
    });
    await hostAdmin.populateRoles();
    await fakeTransaction({
      CollectiveId: collective.id,
      FromCollectiveId: host.id,
      amount: 100000000,
      amountInHostCurrency: 100000000,
      currency: collectiveCurrency,
      hostCurrency: hostCurrency,
    });
    const payoutMethod = await fakePayoutMethod({
      type: PayoutMethodTypes.BANK_ACCOUNT,
      CollectiveId: user.CollectiveId,
      data: payoutData,
    });
    const expense: Expense & { data: any } = await fakeExpense({
      payoutMethod: 'transferwise',
      status: ExpenseStatuses.APPROVED,
      amount: expenseAmount,
      FromCollectiveId: user.CollectiveId,
      CollectiveId: collective.id,
      UserId: user.id,
      currency: expenseCurrency,
      PayoutMethodId: payoutMethod.id,
      type: 'INVOICE',
      description: `Expense in ${expenseCurrency}`,
    });

    await payExpense({ remoteUser: hostAdmin } as any, { id: expense.id });
    await expense.reload();
    expect(expense.status).to.eq(ExpenseStatuses.PROCESSING);
    expect(expense.data.transfer.sourceCurrency).to.eq(host.currency);

    await handleTransferStateChange({
      data: {
        current_state: 'outgoing_payment_sent',
        resource: { id: expense.data.transfer.id, type: 'transfer', profile_id: 1, account_id: 1 },
      },
    } as any);
    await expense.reload();
    const fee = round(expense.data.paymentOption.fee.total * 100);
    const amountInHostCurrency = expense.data.paymentOption.sourceAmount * 100 - fee;

    expect(expense.status).to.eq(ExpenseStatuses.PAID);
    const [debit] = await expense.getTransactions({ where: { type: 'DEBIT' } });
    expect(debit).to.deep.include({
      currency: collectiveCurrency,
      netAmountInCollectiveCurrency: -1 * round((amountInHostCurrency + fee) / debit.hostCurrencyFxRate),
      hostCurrency: hostCurrency,
      amountInHostCurrency: -1 * amountInHostCurrency,
      paymentProcessorFeeInHostCurrency: -1 * fee,
    });

    await snapshotLedger([
      'kind',
      'description',
      'type',
      'amount',
      'paymentProcessorFeeInHostCurrency',
      'amountInHostCurrency',
      'netAmountInCollectiveCurrency',
      'CollectiveId',
      'currency',
      'FromCollectiveId',
      'HostCollectiveId',
      'hostCurrency',
    ]);

    return { expense, debit };
  };

  it('payee.currency = expense.currency = collective.currency = host.currency', async () => {
    const expenseCurrency = 'USD';
    const collectiveCurrency = 'USD';
    const payoutData = {
      accountHolderName: 'Nicolas Cage',
      currency: 'USD',
      type: 'aba',
      details: {
        abartn: '026009593',
        address: {
          city: 'New York',
          state: 'NY',
          country: 'US',
          postCode: '01234',
          firstLine: 'Some Ave',
        },
        legalType: 'PRIVATE',
        accountType: 'CHECKING',
        accountNumber: '12345678',
      },
    };

    await setupTest({
      expenseCurrency,
      collectiveCurrency,
      payoutData,
    });
  });

  it('payee.currency = expense.currency = collective.currency != host.currency', async () => {
    const expenseCurrency = 'EUR';
    const collectiveCurrency = 'EUR';
    const payoutData = {
      type: 'iban',
      details: {
        IBAN: 'FR1420041010050500013M02606',
        address: {
          city: 'Marseille',
          country: 'FR',
          postCode: '13000',
          firstLine: 'xxx',
        },
        legalType: 'PRIVATE',
      },
      currency: 'EUR',
      accountHolderName: 'Le Nicolas Cage',
    };

    await setupTest({
      expenseCurrency,
      collectiveCurrency,
      payoutData,
    });
  });

  it('payee.currency = expense.currency != collective.currency = host.currency', async () => {
    const expenseCurrency = 'EUR';
    const collectiveCurrency = 'USD';
    const payoutData = {
      type: 'iban',
      details: {
        IBAN: 'FR1420041010050500013M02606',
        address: {
          city: 'Marseille',
          country: 'FR',
          postCode: '13000',
          firstLine: 'xxx',
        },
        legalType: 'PRIVATE',
      },
      currency: 'EUR',
      accountHolderName: 'Le Nicolas Cage',
    };

    await setupTest({
      expenseCurrency,
      collectiveCurrency,
      payoutData,
    });
  });

  it('payee.currency != expense.currency = collective.currency = host.currency', async () => {
    const expenseCurrency = 'USD';
    const collectiveCurrency = 'USD';
    const payoutData = {
      type: 'iban',
      details: {
        IBAN: 'FR1420041010050500013M02606',
        address: {
          city: 'Marseille',
          country: 'FR',
          postCode: '13000',
          firstLine: 'xxx',
        },
        legalType: 'PRIVATE',
      },
      currency: 'EUR',
      accountHolderName: 'Le Nicolas Cage',
    };

    await setupTest({
      expenseCurrency,
      collectiveCurrency,
      payoutData,
    });
  });

  it('payee.currency != expense.currency = collective.currency != host.currency', async () => {
    const expenseCurrency = 'EUR';
    const collectiveCurrency = 'EUR';
    const payoutData = {
      type: 'sort_code',
      details: {
        address: {
          city: 'London',
          country: 'GB',
          postCode: 'ODF 6DB',
          firstLine: '7 Road',
        },
        legalType: 'PRIVATE',
        sortCode: '231470',
        accountNumber: '28821822',
      },
      currency: 'GBP',
      accountHolderName: 'Sir Nicolas Cage',
    };

    await setupTest({
      expenseCurrency,
      collectiveCurrency,
      payoutData,
    });
  });

  it('host.currency = payee.currency != expense.currency = collective.currency', async () => {
    const expenseCurrency = 'EUR';
    const collectiveCurrency = 'EUR';
    const payoutData = {
      accountHolderName: 'Nicolas Cage',
      currency: 'USD',
      type: 'aba',
      details: {
        abartn: '026009593',
        address: {
          city: 'New York',
          state: 'NY',
          country: 'US',
          postCode: '01234',
          firstLine: 'Some Ave',
        },
        legalType: 'PRIVATE',
        accountType: 'CHECKING',
        accountNumber: '12345678',
      },
    };

    const { expense, debit } = await setupTest({
      expenseCurrency,
      collectiveCurrency,
      payoutData,
    });

    expect(debit.hostCurrencyFxRate).to.be.gt(1);
    expect(debit.amountInHostCurrency).to.be.eq(-1 * round(expense.amount * debit.hostCurrencyFxRate));
  });
});
