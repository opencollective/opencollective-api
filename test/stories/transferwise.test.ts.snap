// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/test/stories/transferwise.test.ts host.currency = payee.currency != expense.currency = collective.currency 1`] = `
"
| kind    | description    | type   | amount    | paymentFee | amountInHostCurrency | netAmountInCollectiveCurrency | To          | currency | From                   | Host                   | hostCurrency |
| ------- | -------------- | ------ | --------- | ---------- | -------------------- | ----------------------------- | ----------- | -------- | ---------------------- | ---------------------- | ------------ |
| EXPENSE | Expense in EUR | CREDIT | 10099     | -105       | 10761                | 10000                         | Donnortello | EUR      | Babel EUR              | NULL                   | USD          |
| EXPENSE | Expense in EUR | DEBIT  | -10000    | -105       | -10656               | -10099                        | Babel EUR   | EUR      | Donnortello            | Open Source Collective | USD          |
|         |                | CREDIT | 100000000 | 0          | 100000000            | 100000000                     | Babel EUR   | EUR      | Open Source Collective | NULL                   | USD          |"
`;

exports[`/test/stories/transferwise.test.ts payee.currency != expense.currency = collective.currency != host.currency 1`] = `
"
| kind    | description    | type   | amount    | paymentFee | amountInHostCurrency | netAmountInCollectiveCurrency | To          | currency | From                   | Host                   | hostCurrency |
| ------- | -------------- | ------ | --------- | ---------- | -------------------- | ----------------------------- | ----------- | -------- | ---------------------- | ---------------------- | ------------ |
| EXPENSE | Expense in EUR | CREDIT | 10080     | -85        | 10742                | 10000                         | Donnortello | EUR      | Babel EUR              | NULL                   | USD          |
| EXPENSE | Expense in EUR | DEBIT  | -10000    | -85        | -10657               | -10080                        | Babel EUR   | EUR      | Donnortello            | Open Source Collective | USD          |
|         |                | CREDIT | 100000000 | 0          | 100000000            | 100000000                     | Babel EUR   | EUR      | Open Source Collective | NULL                   | USD          |"
`;

exports[`/test/stories/transferwise.test.ts payee.currency != expense.currency = collective.currency = host.currency 1`] = `
"
| kind    | description    | type   | amount    | paymentFee | amountInHostCurrency | netAmountInCollectiveCurrency | To          | currency | From                   | Host                   | hostCurrency |
| ------- | -------------- | ------ | --------- | ---------- | -------------------- | ----------------------------- | ----------- | -------- | ---------------------- | ---------------------- | ------------ |
| EXPENSE | Expense in USD | CREDIT | 10080     | -80        | 10080                | 10000                         | Donnortello | USD      | Babel USD              | NULL                   | USD          |
| EXPENSE | Expense in USD | DEBIT  | -10000    | -80        | -10000               | -10080                        | Babel USD   | USD      | Donnortello            | Open Source Collective | USD          |
|         |                | CREDIT | 100000000 | 0          | 100000000            | 100000000                     | Babel USD   | USD      | Open Source Collective | NULL                   | USD          |"
`;

exports[`/test/stories/transferwise.test.ts payee.currency = expense.currency != collective.currency = host.currency 1`] = `
"
| kind    | description    | type   | amount    | paymentFee | amountInHostCurrency | netAmountInCollectiveCurrency | To          | currency | From                   | Host                   | hostCurrency |
| ------- | -------------- | ------ | --------- | ---------- | -------------------- | ----------------------------- | ----------- | -------- | ---------------------- | ---------------------- | ------------ |
| EXPENSE | Expense in EUR | CREDIT | 10739     | -83        | 10739                | 10656                         | Donnortello | USD      | Babel USD              | NULL                   | USD          |
| EXPENSE | Expense in EUR | DEBIT  | -10656    | -83        | -10656               | -10739                        | Babel USD   | USD      | Donnortello            | Open Source Collective | USD          |
|         |                | CREDIT | 100000000 | 0          | 100000000            | 100000000                     | Babel USD   | USD      | Open Source Collective | NULL                   | USD          |"
`;

exports[`/test/stories/transferwise.test.ts payee.currency = expense.currency = collective.currency != host.currency 1`] = `
"
| kind    | description    | type   | amount    | paymentFee | amountInHostCurrency | netAmountInCollectiveCurrency | To          | currency | From                   | Host                   | hostCurrency |
| ------- | -------------- | ------ | --------- | ---------- | -------------------- | ----------------------------- | ----------- | -------- | ---------------------- | ---------------------- | ------------ |
| EXPENSE | Expense in EUR | CREDIT | 10078     | -83        | 10739                | 10000                         | Donnortello | EUR      | Babel EUR              | NULL                   | USD          |
| EXPENSE | Expense in EUR | DEBIT  | -10000    | -83        | -10656               | -10078                        | Babel EUR   | EUR      | Donnortello            | Open Source Collective | USD          |
|         |                | CREDIT | 100000000 | 0          | 100000000            | 100000000                     | Babel EUR   | EUR      | Open Source Collective | NULL                   | USD          |"
`;

exports[`/test/stories/transferwise.test.ts payee.currency = expense.currency = collective.currency = host.currency 1`] = `
"
| kind    | description    | type   | amount    | paymentFee | amountInHostCurrency | netAmountInCollectiveCurrency | To          | currency | From                   | Host                   | hostCurrency |
| ------- | -------------- | ------ | --------- | ---------- | -------------------- | ----------------------------- | ----------- | -------- | ---------------------- | ---------------------- | ------------ |
| EXPENSE | Expense in USD | CREDIT | 10105     | -105       | 10105                | 10000                         | Donnortello | USD      | Babel USD              | NULL                   | USD          |
| EXPENSE | Expense in USD | DEBIT  | -10000    | -105       | -10000               | -10105                        | Babel USD   | USD      | Donnortello            | Open Source Collective | USD          |
|         |                | CREDIT | 100000000 | 0          | 100000000            | 100000000                     | Babel USD   | USD      | Open Source Collective | NULL                   | USD          |"
`;
